typing/typecore.ml

```diff
@@ -664,19 +685,40 @@ and build_as_type_aux env p =
           unify_pat env p ty_res'
         end in
       Array.iter do_label lbl.lbl_all;
-      ty
+      ty, p.pat_mode
   | Tpat_or(p1, p2, row) ->
       begin match row with
         None ->
-          let ty1 = build_as_type env p1 and ty2 = build_as_type env p2 in
+          let ty1, mode1 = build_as_type_and_mode env p1 in
+          let ty2, mode2 = build_as_type_and_mode env p2 in
           unify_pat env {p2 with pat_type = ty2} ty1;
-          ty1
+          ty1, Value_mode.join [mode1; mode2]
       | Some row ->
           let row = row_repr row in
-          newty (Tvariant{row with row_closed=false; row_more=newvar()})
+          let all_constant =
+            List.for_all
+              (function
+                | _, (Rpresent (Some _) | Reither (false, _, _, _)) -> false
+                | _ -> true)
+              row.row_fields
+          in
+          let mode =
+            if all_constant then Value_mode.newvar ()
+            else p.pat_mode
+          in
+          let ty =
+            newty (Tvariant{row with row_closed=false; row_more=newvar()})
+          in
+          ty, mode
       end
-  | Tpat_any | Tpat_var _ | Tpat_constant _
-  | Tpat_array _ | Tpat_lazy _ -> p.pat_type
+  | Tpat_constant cst ->
+      let mode =
+        if is_allocated_constant cst then p.pat_mode
```

Comment:
The logic is correct here, but since I think `is_allocated_constant` should be removed from #37 I'd probably spell it `is_immediate p.pat_type` or similar.